% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/clonalQuant.R
\name{clonalQuant}
\alias{clonalQuant}
\title{Quantify the unique clones by group or sample}
\usage{
clonalQuant(
  input.data,
  cloneCall = "strict",
  chain = "both",
  scale = FALSE,
  group.by = NULL,
  order.by = NULL,
  exportTable = FALSE,
  palette = "inferno"
)
}
\arguments{
\item{input.data}{The product of \code{\link[=combineTCR]{combineTCR()}},
\code{\link[=combineBCR]{combineBCR()}}, or \code{\link[=combineExpression]{combineExpression()}}.}

\item{cloneCall}{How to call the clone - VDJC gene (\strong{gene}),
CDR3 nucleotide (\strong{nt}), CDR3 amino acid (\strong{aa}),
VDJC gene + CDR3 nucleotide (\strong{strict}) or a custom variable
in the data}

\item{chain}{indicate if both or a specific chain should be used -
e.g. "both", "TRA", "TRG", "IGH", "IGL"}

\item{scale}{Converts the graphs into percentage of unique clones}

\item{group.by}{The column header used for grouping}

\item{order.by}{A vector of specific plotting order or "alphanumeric"
to plot groups in order}

\item{exportTable}{Returns the data frame used for forming the graph}

\item{palette}{Colors to use in visualization - input any
\link[grDevices:palettes]{hcl.pals}}
}
\value{
ggplot of the total or relative unique clones
}
\description{
This function quantifies unique clones. The unique clones
can be either reported as a raw output or scaled to the total number of
clones recovered using the scale parameter.
}
\examples{
#Making combined contig data
combined <- combineTCR(contig_list, 
                        samples = c("P17B", "P17L", "P18B", "P18L", 
                                    "P19B","P19L", "P20B", "P20L"))
clonalQuant(combined, 
            cloneCall="strict", 
            scale = TRUE)

}
\concept{Visualizing_Clones}
