% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/loadContigs.R
\name{loadContigs}
\alias{loadContigs}
\title{Loading the contigs derived from single-cell sequencing}
\usage{
loadContigs(input, format = "10X")
}
\arguments{
\item{input}{The directory in which contigs are located or a list with contig
elements}

\item{format}{The format of the single-cell contig, currently supporting:
"10X", "AIRR", "BD", "Dandelion", "JSON", "MiXCR", "ParseBio", "Omniscope",
"TRUST4", "WAT3R", and "Immcantation"}
}
\value{
List of contigs for compatibility  with \code{\link[=combineTCR]{combineTCR()}} or
\code{\link[=combineBCR]{combineBCR()}}
}
\description{
This function generates a contig list and formats the data to allow for
function with  \code{\link[=combineTCR]{combineTCR()}} or \code{\link[=combineBCR]{combineBCR()}}. If
using data derived from filtered outputs of 10X Genomics, there is no
need to use this function as the data is already compatible.
}
\details{
The files that this function parses includes:
\itemize{
\item 10X =  "filtered_contig_annotations.csv"
\item AIRR = "airr_rearrangement.tsv"
\item BD = "Contigs_AIRR.tsv"
\item Dandelion = "all_contig_dandelion.tsv"
\item Immcantation = "data.tsv"
\item JSON = ".json"
\item ParseBio = "barcode_report.tsv"
\item MiXCR = "clones.tsv"
\item Omniscope = ".csv"
\item TRUST4 = "barcode_report.tsv"
\item WAT3R = "barcode_results.csv"
}
}
\examples{
TRUST4 <- read.csv("https://www.borch.dev/uploads/contigs/TRUST4_contigs.csv")
contig.list <- loadContigs(TRUST4, format = "TRUST4")

BD <- read.csv("https://www.borch.dev/uploads/contigs/BD_contigs.csv")
contig.list <- loadContigs(BD, format = "BD")

WAT3R <- read.csv("https://www.borch.dev/uploads/contigs/WAT3R_contigs.csv")
contig.list <- loadContigs(WAT3R, format = "WAT3R")

}
\concept{Loading_and_Processing_Contigs}
